### PR TITLE
Fix/repair existing unit tests

### PR DESCRIPTION
**Pull request by henrijaeger**
- [x] Repair unit tests for ItemController
- [x] Repair unit tests for Item
- [x] Add some(!) tests for LocationController
- [ ] Add edge cases and invalid inputs => more test cases!
- [ ] Discuss: How (or even if) to test methods that use Firebase listeners (onSuccess/onFailure/snapshotListener)

### CODE REVIEW

**Comment by henrijaeger** [comment]
@PassiDel @frcroth @AlexTerhar: Could you take a look at the test cases I've created?
Do you have any comments on them?
Of course, some of them have to be extended by unexpected input values.
Unfortunately, I did not manage to test methods, that use Firebase listeners (onSuccessListener / onFailureListener / snapshotListener), such as overrideItem, addItem and many other methods, especially in UserController. 
How do we test those methods? 
Are we even supposed to test them?

File: app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt [hunk 1]
```diff
@@ -1,154 +1,129 @@
...
-    fun testParsingWithoutSharedEmail() {
-        val modifiedItemData = mapWithNullValue(itemData, "shared_email")
-        val parsedItem = ItemController.parse(modifiedItemData, firebaseId)
-        val itemExpected = Item(firebaseId, name, description, isShared, quantity.toInt(), unit, null, bestByDate, location, geohash, contactName, contactEmail, ownerRef)
-        assertEquals(itemExpected, parsedItem)
+    fun testSaveItem_newItem() {
+        mockkStatic(FirebaseFirestore::class)
+        every { FirebaseFirestore.getInstance() } returns mockk(relaxed = true)
+        val dr = Mockito.mock(DocumentReference::class.java)
+        val foo = spyk(ItemController, recordPrivateCalls = true)
```
Comment (frcroth) [inline app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt:90]: rename variable

File: app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt [hunk 2]
```diff
@@ -1,154 +1,129 @@
...
     }
 
     @Test
-    fun testParsingWithoutIsShared() {
-        val modifiedItemData = mapWithNullValue(itemData, "is_shared")
-        val parsedItem = ItemController.parse(modifiedItemData, firebaseId)
-        val itemExpected = Item(firebaseId, name, description, null, quantity.toInt(), unit, sharedEmail, bestByDate, location, geohash, contactName, contactEmail, ownerRef)
-        assertEquals(itemExpected, parsedItem)
-    }
+    fun testParsing_withoutItemOwner() {
```
Comment (frcroth) [inline app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt:73]: Method name does not conform to standard

File: app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt [hunk 3]
```diff
@@ -1,154 +1,129 @@
...
+        verify(exactly = 1) { itemController["overrideItem"](any<Item>(), any<() -> kotlin.Unit>(), any<(Exception) -> kotlin.Unit>()) }
     }
 
-    private fun mapWithNullValue(mutableMap: Map<String, Any>, field: String): MutableMap<String, Any> {
-        var modifiedHashMap = HashMap<String, Any?>()
-        for ((key, value) in mutableMap) {
-            if (key == field) modifiedHashMap[key] = null
-            else modifiedHashMap[key] = value
+    private fun setUpDocumentSnapshot(dc: DocumentSnapshot, values: Map<String, Any>, id: String) {
+        // Mockito works, by intercepting method calls.
```
Comment (frcroth) [inline app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt:109]:
```suggestion
        // Mockito works by intercepting method calls.
```

**Review by frcroth — COMMENTED** [review]
I think it goes very much in the right direction, these tests make sense

File: app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt [hunk 4]
```diff
@@ -1,154 +1,129 @@
...
 
-    @Test
-    fun testParsingWithoutName() {
-        val modifiedItemData = mapWithNullValue(itemData, "name")
-        val parsedItem = ItemController.parse(modifiedItemData, firebaseId)
-        val itemExpected = Item(firebaseId, null, description, isShared, quantity.toInt(), unit, sharedEmail, bestByDate, location, geohash, contactName, contactEmail, ownerRef)
-        assertEquals(itemExpected, parsedItem)
+    @After
+    fun tearDown() {
+        setUpDocumentSnapshot(itemDocSnap, dummyValues, dummyFirebaseId)
```
Comment (PassiDel) [inline app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt:56]: Why is this called on tearDown?

File: app/src/androidTest/java/app/wefridge/wefridge/LocationControllerUnitTest.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import org.junit.runner.RunWith
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.junit.Before
+
+
+/*
+* NOTE: Run the following tests with device language set to English.
+* */
+@RunWith(AndroidJUnit4::class)
+class LocationControllerUnitTest {
```
Comment (PassiDel) [inline app/src/androidTest/java/app/wefridge/wefridge/LocationControllerUnitTest.kt:18]: Does it make sense for us to test the Android GeoCoder? We didn't make it ourselves, so we have correspondingly little control.

**Review by PassiDel — APPROVED** [review]
Looks fine to me. I think we should merge this PR (even prematurely) to continue working with the LocationController (I'm still waiting with #57).

> Unfortunately, I did not manage to test methods, that use Firebase listeners (onSuccessListener / onFailureListener / snapshotListener), such as getItemsSnapshot and many other methods in UserController.
> How do we test those methods?
> Are we even supposed to test them?

I think Andreas said, we don't have to test these firebase methods, rather our logic. Would be too many variables that we don't control.

File: app/src/androidTest/java/app/wefridge/wefridge/LocationControllerUnitTest.kt (hunk 5 above)
Comment (henrijaeger) [reply app/src/androidTest/java/app/wefridge/wefridge/LocationControllerUnitTest.kt:18]: That's right. Shall I remove the tests for that? @frcroth what is your opinion on that?

Comment (frcroth) [reply app/src/androidTest/java/app/wefridge/wefridge/LocationControllerUnitTest.kt:18]: You mean the whole file? No I think it can stay

File: app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt (hunk 4 above)
Comment (henrijaeger) [reply app/src/test/java/app/wefridge/wefridge/ItemControllerUnitTest.kt]: Calling it in the setUp method would have caused 'testParsingWithoutItemOwner' to fail. I have refactored "setUpDocumentSnapshot" now, so that we get can call a modfied version of it in the setUp method.

**Comment by henrijaeger** [comment]
> Looks fine to me. I think we should merge this PR (even prematurely) to continue working with the LocationController (I'm still waiting with #57).

Isn't LocationController already available in main?

File: app/src/androidTest/java/app/wefridge/wefridge/LocationControllerUnitTest.kt (hunk 5 above)
Comment (henrijaeger) [reply app/src/androidTest/java/app/wefridge/wefridge/LocationControllerUnitTest.kt:18]: Okay, then I will leave the file in the project for now.

**Comment by henrijaeger** [comment]
Merge repaired and new unit tests into main.
Add further unit tests in a new branch.

**Comment by PassiDel** [comment]
> > Looks fine to me. I think we should merge this PR (even prematurely) to continue working with the LocationController (I'm still waiting with #57).
> 
> Isn't LocationController already available in main? Anyways, I can merge this pull request and will then continue working on further tests.

Yes but you moved two methods into the companion object, I didn't realize that I didn't need them. :sweat_smile:
